### PR TITLE
Add Exception Details to RPC Exceptions

### PR DESCRIPTION
**Pull request by Warfields**
Propagates details from exceptions over RPC.

### CODE REVIEW

File: c++/src/kj/exception.h [hunk 1]
```diff
@@ -135,8 +135,14 @@ class Exception {
...
 
   using DetailTypeId = unsigned long long;
+  struct Detail {
+    DetailTypeId id;
+    kj::Array<byte> value;
+  };
 
   kj::Maybe<kj::ArrayPtr<const byte>> getDetail(DetailTypeId typeId) const;
+  kj::ArrayPtr<const Detail> getDetails() const;
+  bool hasDetails() const;
```
Comment (jasnell) [inline c++/src/kj/exception.h:145]: `hasDetails()` isn't strictly necessary here. Someone could just check `getDetails().size() != 0`, or have `getDetail()` return a `kj::Maybe<kj::ArrayPtr<const Detail>>` (but I prefer just the size check).

Comment (Warfields) [reply c++/src/kj/exception.h:145]: Done, I was over complicating that.

Comment (Warfields) [reply c++/src/kj/exception.h]: Helps if I push up the change :D

**Review by jasnell — COMMENTED** [review]
The changes to `kj::Exception` LGTM. I'll leave it to someone else to review the `src/capnp/rpc*` changes

File: c++/src/capnp/test-util.c++ [hunk 2]
```diff
@@ -1192,6 +1192,13 @@ kj::Promise<void> TestMoreStuffImpl::throwRemoteException(ThrowRemoteExceptionCo
   return KJ_EXCEPTION(FAILED, "remote exception: test exception");
 }
 
+kj::Promise<void> TestMoreStuffImpl::throwExceptionWithDetail(
+    ThrowExceptionWithDetailContext context) {
+  auto e = KJ_EXCEPTION(FAILED, "test exception");
+  e.setDetail(1, kj::heapArray<byte>(0));
```
Comment (kentonv) [inline c++/src/capnp/test-util.c++:1198]: We should probably test a non-empty detail to make sure the bytes are in fact encoded correctly?

**Pull request merged**